### PR TITLE
Location getter + abortable check

### PR DESCRIPTION
**Pull request by daun**
**Description**

- Re-implement some changes from the `visit:abort` branch that make sense

**Location object**
- Store the current location as a `Location` object at `swup.location`
- Allows checking the current hash in addition to just the path/query
- Save the hash of the previous page in the visit object
- Deprecate `swup.currentPageUrl` in favor of `swup.location.url`

```js
if (swup.location.url === '/about') {}
if (swup.location.hash) {}
if (visit.from.hash === '#detail') {}
```

**Location update**
- Update browser url at earliest possible time

**Aborting visits**
- Add `visit.abortable` getter that defines when a visit can be aborted
- Visits that have started entering cannot be aborted
- ~~History visits cannot be aborted~~ (wrong assumption, needs to be abortable)
- Refactor aborting visits from `swup.navigate()`

**Note to self**
- Not allowing aborting history doesn't make too much sense here (we only abort visits if there's a new one)

**Checks**

- [x] The PR is submitted to the `master` branch
- [x] The code was linted before pushing (`npm run lint`)
- [x] All tests are passing (`npm run test`)
- [x] New or updated tests are included
- [ ] The documentation was updated as required

### CODE REVIEW

**Comment by hirasso** [comment]
> Not allowing aborting history doesn't make too much sense here (we only abort visits if there's a new one)

It should read "allowing", not "Not allowing", right? :)

**Comment by daun** [comment]
@hirasso No, I think that's what I actually meant to write. In the following sense: by not allowing history visits to be aborted, we now make it impossible to supersede history visits. That would be the case when clicking the back button a few times in quick succession. If the cache is disabled or history visits are animated, every one of these history visits will have to finish completely before queueing the next one. Hence, we should allow aborting history visits if there's a new visit. I need to remove that part from this PR.

The really not allowing to abort them was only necessary in the previous PR where one could abort a visit from anywhere at any point. But even there we should have added a special condition for quick history visits in succession.

This is all very confusing :/ I really wish we had a nice state machine 🤖

**Comment by hirasso** [comment]
It's indeed pretty confusing... thank you for clarifying what exactly you meant.

**Comment by daun** [comment]
Addressed the open points and fixed the tests. Ready for review :)

File: src/modules/navigate.ts [hunk 1]
```diff
@@ -123,8 +117,21 @@ export async function performNavigation(
 
 	try {
 		await this.hooks.call('visit:start', visit, undefined);
+
 		visit.state = VisitState.STARTED;
 
+		// Create/update history record if this is not a popstate call or leads to the same URL
```
Comment (hirasso) [inline src/modules/navigate.ts:123]: Moving the bit where the URL updates feels pretty scary to me. I can recall various situations where we discovered quite late in the process that the exact moment where the URL was updated had a lot of implications.

Why did you move it?

**Comment by daun** [comment]
This is doing a bit too much — slimmed down PR incoming :)

**Pull request closed without merging**